maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,320 @@
+
+        for pm in self._machines:
+            total_util_cpu: int = 0
+            for vm_id in pm.vm_set:
+                vm = self._live_vm[vm_id]
+                total_util_cpu += vm.util_cpu * vm.req_cpu / 100
+            util_cpu = total_util_cpu / pm.cap_cpu * 100
+            pm.update_util(tick=self._tick, util_cpu=util_cpu)
+            self._energy_consumption += self._cpu_util_to_energy_consumption(pm.util_cpu)
+
+    def _cpu_util_to_energy_consumption(self, cpu_util: float) -> float:
+        """Convert the CPU utilization to energy consumption."""
+
+        power: float = self._conf["calibration parameter"]
+        # NOTE: Energy comsumption parameters should refer to more research.
+        busy_power = self._conf["busy_power"]
+        idle_power = self._conf["idle_power"]
+        return idle_power + (busy_power - idle_power) * (2 * cpu_util - pow(cpu_util, power))
+
+    def _postpone_vm_requirement(self, postpone_type: PostponeType, vm_id: int, remaining_buffer_time: int):
+        """Postpone VM requirement."""
+
+        if remaining_buffer_time >= self._delay_duration:
+            if postpone_type == PostponeType.Resource:
+                self._total_latency.latency_due_to_resource += self._delay_duration
+            elif postpone_type == PostponeType.Agent:
+                self._total_latency.latency_due_to_agent += self._delay_duration
+
+            postpone_payload = self._pending_vm_req_payload[vm_id]
+            postpone_payload.remaining_buffer_time -= self._delay_duration
+            postpone_event = self._event_buffer.gen_cascade_event(
+                tick=self._tick + self._delay_duration,
+                event_type=Events.REQUIREMENTS,
+                payload=postpone_payload
+            )
+            self._event_buffer.insert_event(event=postpone_event)
+        else:
+            # Fail
+            # Pop out VM requirement payload.
+            self._pending_vm_req_payload.pop(vm_id)
+            # Add failed requirements.
+            self._failed_requirements += 1
+
+    def _get_valid_pm(self, vm_req_cpu: int) -> List[ValidPhysicalMachine]:
+        """Check all valid PMs.
+
+        There are three situations:
+            1. All PMs are empty.
+            2. Some PMs are empty but others are not.
+            3. All PMs are not empty.
+        Situation 1: Return the first empty PM info.
+        Situation 2: Return all PMs with enough resources but not empty and plus the first empty PM.
+        Situation 3: Return all PMs with enough resources but not empty.
+
+        Args: vm_req_cpu (int): The CPU resource of the VM requirement.
+        """
+        # NOTE: Should we implement this logic inside the action scope?
+        # TODO: In oversubscribable scenario, we should consider more situations, like
+        #       the PM type (oversubscribable and non-oversubscribable).
+
+        valid_pm_list = []
+        for pm in self._machines:
+            if pm.req_cpu == 0:
```

Comment:
what's the meaning of the pm CPU requirement?

Reply:
It means all VMs' (on that PM) CPU requirement.

Reply:
maybe use, pm_cpu_allocation